src/core/hle/service/fs/file.cpp

```diff
@@ -76,22 +75,96 @@ void File::Read(Kernel::HLERequestContext& ctx) {
                   offset, length, backend->GetSize());
     }
 
-    IPC::RequestBuilder rb = rp.MakeBuilder(2, 2);
+    if (backend->AllowsCachedReads()) {
+        struct AsyncData {
+            // Input
+            u32 length;
+            u64 offset;
+            std::chrono::steady_clock::time_point pre_timer;
+            bool cache_ready;
+
+            // Output
+            ResultCode ret{0};
+            Kernel::MappedBuffer* buffer;
+            std::unique_ptr<u8*> data;
+            size_t read_size;
+        };
+
+        auto async_data = std::make_shared<AsyncData>();
+        async_data->buffer = &rp.PopMappedBuffer();
+        async_data->length = length;
+        async_data->offset = offset;
+        async_data->cache_ready = backend->CacheReady(offset, length);
+        if (!async_data->cache_ready) {
+            async_data->pre_timer = std::chrono::steady_clock::now();
+        }
+
+        // LOG_DEBUG(Service_FS, "cache={}, offset={}, length={}", cache_ready, offset, length);
+        ctx.RunAsync(
+            [this, async_data](Kernel::HLERequestContext& ctx) {
+                async_data->data =
+                    std::make_unique<u8*>(static_cast<u8*>(operator new(async_data->length)));
+                ResultVal<std::size_t> read =
+                    backend->Read(async_data->offset, async_data->length, *async_data->data);
```

Comment:
```suggestion
                const auto read = backend->Read(async_data->offset, async_data->length, *async_data->data);
```
Since data below also uses auto, for consistency